airflow/contrib/hooks/gcp_sql_hook.py

```diff
@@ -261,3 +280,559 @@ def _wait_for_operation_to_complete(self, project_id, operation_name):
+                "unix_socket={socket_path}"
+        },
+        "public": {
+            "ssl":
+                "mysql://{user}:{password}@{public_ip}:{public_port}/{database}?"
+                "ssl={ssl_spec}",
+            "non-ssl":
+                "mysql://{user}:{password}@{public_ip}:{public_port}/{database}"
+        }
+    }
+}
+
+CLOUD_SQL_VALID_DATABASE_TYPES = ['postgres', 'mysql']
+
+
+# noinspection PyAbstractClass
+class CloudSqlDatabaseHook(BaseHook):
+    """Serves DB connection configuration for CloudSQL (Connections
+    of *cloudsql://* type).
+
+    The hook is a "meta" one - it does not perform an actual connection,
+    it is there to retrieve all the parameters configured in cloudsql:// connection,
+    start/stop Cloud Sql Proxy if needed, dynamically generate Postgres or MySQL
+    connection in the database and return an actual Postgres or MySQL hook.
+    The returned Postgres/MySQL hooks are using direct connection or Cloud Sql
+    Proxy socket/tcp as configured.
+
+    Main parameters of the hook are retrieved from the standard URI components:
+
+    * **user** - User name to authenticate to the database (from login of the URI).
+    * **password** - Password to authenticate to the database (from password of the URI)
+    * **public_ip** - IP to connect to for public connection (from host of the URI)
+    * **public_port** - Port to connect to for public connection (from port of the URI)
+    * **database** - Database to connect to (from schema of the URI)
+
+    Remaining parameters are retrieved from the extras (URI query parameters):
+
+    * **project_id** - Google Cloud Platform project where the Cloud SQL instance exists.
+    * **instance** -  Name of the instance of the Cloud SQL database instance.
+    * **location** - The location of the cloud sql instance (for example europe-west1).
+    * **database_type** - The type of the database instance (mysql or postgres).
+    * **use_proxy** - (default False) Whether SQL proxy should be used to connect to Cloud
+      SQL DB.
+    * **use_ssl** - (default False) Whether SSL should be used to connect to Cloud SQL DB.
+      You cannot use proxy and ssl together.
+    * **sql_proxy_use_tcp** - (default False) If set to true, TCP is used to connect via
+      proxy, otherwise UNIX sockets are used.
+    * **sql_proxy_binary_path** - Optional path to sql proxy binary. If the binary is not
+      specified or the binary is not present, it is automatically downloaded.
+    * **sql_proxy_version** -  Specific version of the proxy to download (for example
+      v1.13). If not specified, the latest version is downloaded.
+    * **sslcert** - Path to client certificate to authenticate when SSL is used.
+    * **sslkey** - Path to client private key to authenticate when SSL is used.
+    * **sslrootcert** - Path to server's certificate to authenticate when SSL is used.
+    """
+    _conn = None
+
+    def __init__(self, cloudsql_conn_id='google_cloud_sql_default'):
```

Comment:
Can we rename `cloudsql_conn_id ` to `gcp_ cloudsql_conn_id` as it is a gcp product